rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,179 @@
+with. For example, if the metric was published two dimensions,
+availability-zone and load balancer name, then both dimensions must be
+specified to retrieve the metric. CloudWatch won't automatically aggregate
+across the other if only one is specified.
+
+This source will use the default AWS credential chain similar to the
+`aws_cloudwatch_metrics` sink.
+
+The output (shown here in prometheus format) will publish metrics in the form
+of:
+
+```
+<namespace>_<metric_name>_<metric_stat>{<dimension_key>=<dimension_value>*}
+```
+
+All metrics will be additionally tagged with the region.
+
+For example:
+
+```
+aws_ec2_cpu_credit_balance_average{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_maximum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_minimum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+aws_ec2_cpu_credit_balance_sample_count{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 1
+aws_ec2_cpu_credit_balance_sum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+```
+
+I think we will want to provide example configs that collect common metrics for
+various namespaces (like EC2, S3, etc.).
+
+## Rationale
+
+Broadening AWS platform support.
+
+Otherwise, users will need to run `telegraf` or another agent to ingest this
+data.
+
+## Prior Art
+
+- [telegraf](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/cloudwatch)
+- [prometheus exporter](https://github.com/prometheus/cloudwatch_exporter)
+- [filebeat](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-cloudwatch.html)
+
+## Drawbacks
+
+The additional maintenance burden of a new source.
+
+## Alternatives
+
+There is alternate API,
+[`GetMetricStatistics'](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricStatistics.html)
+that we could use, but I think it will result in less overall throughput due to
+API rate limit restrictions:
+
+`GetMetricStatistics`
+
+* Allows 400 transactions / sec
+* Each transaction can query the 5 standard metrics and up to 10 extended
+  metrics for one metric name / dimension combination
+* This should allow for: `400 * 5 = 2000` standard metrics / second
+
+`GetMetricData`
+
+* Allows 50 transactions / sec
+* Each transaction can query up to 500 metric name / metric stat (like
+  `average`) / dimension combination
+* This should allow for: `50 * 500 = 25000` standard metrics / second
+
+These are theoretical maxes. They could differ depending on how many metric
+statistics are queried for each metric. Querying less than the 5 standard ones
+would further advantage `GetMetricData`.
+
+## Outstanding Questions
+
+- Do we care about providing a different rate limit for refreshing available
+	metrics? It uses a different API (`ListMetrics`) with different limits than
+	`GetMetricData`.
+- Do we want to allow multiple regions to be specified at once? If so, the
```

Comment:
We don't currently do this anywhere else, and I'm wondering how this would work with authentication and environment variables like `AWS_REGION`? So my vote is no.

Reply:
Definitely no. We shoudldn't, even if we could via auth, cross security boundaries.

Reply:
I think we don't currently allow multiple regions anywhere else because the only other AWS integrations we have are sinks, where it wouldn't make sense. This is the first source. We could only fallback to `AWS_REGION` if the regions are not specified in the config.

The problem I'm thinking this would help solve is for companies that deploy the same infrastructure / services in multiple AWS regions for availability purposes. They will likely want to collect the same metrics from all of these regions. If we don't allow specification of multiple, they will likely need to copy/paste the config and keep them all in-sync. This problem could also be solved in other ways: like generating vector configs from another format or maybe the macro idea.

Consider that, for example, the Datadog AWS integration just pulls all metrics from all regions using the same AWS IAM role.

For authentication, in the multi-region scenario, typically the AWS IAM user or role being used has access to the same resources across regions. If they don't, then we'll get an AWS authentication error, which we can report.

@jamtur01 I don't typically think of regions as security boundaries, could you elaborate on that? When I think of security boundaries in AWS I usually think of separate accounts or using IAM policies to restrict.

Reply:
Datadog does this? If so, then I'm a fan.

Reply:
I've found most folks operate independently in a single region and use multi-region for redundancy/geo failover. So monitoring tends to be single region-centric (or multiple monitoring instances across multiple regions) so I tend to think about regions as boundaries. I realize that is probably my bias. So I defer my objection in favor of the DD approach. :)

Reply:
:+1: I thought about this some more and feel like users would probably actually typically want to run one `vector` aggregator per region in the scenario I suggested (for availability and latency reasons), but I think it could still be useful for users that run just one in some sort of centralized monitoring region and I'm hopeful it doesn't introduce too much complexity so we can give it a shot.